From 244a8b67824e9dfd46f2bfe83bb6714d85f9b581 Mon Sep 17 00:00:00 2001
From: Roman Stratiienko <r.stratiienko@gmail.com>
Date: Sun, 27 Mar 2022 15:16:38 +0300
Subject: [PATCH] GLODROID: BT: HCI: HACK: Spoof status for
 WriteSecureConnectionsHostSupport

rtl8723cs bluetooth will return NOT_SUPPORTED(data[5]==1) for the
WriteSecureConnectionsHostSupport (0xC7A) requests, which will fail assert HERE [1]
Replace status with SUCCESS(data[5]=0) to bypass the assert.
Bluetooth functionality is limited for the PP.

[1]: https://cs.android.com/android/platform/superproject/+/master:packages/modules/Bluetooth/system/gd/hci/controller.cc;l=225;drc=fe2308d15f93002e9dfb9a5f04d8178800ec83ea
Signed-off-by: Roman Stratiienko <r.stratiienko@gmail.com>
---
 net/bluetooth/hci_core.c | 21 +++++++++++++++++++++
 1 file changed, 21 insertions(+)

diff --git a/net/bluetooth/hci_core.c b/net/bluetooth/hci_core.c
index cabe8eb4c14f..7d0c12dc2094 100644
--- a/net/bluetooth/hci_core.c
+++ b/net/bluetooth/hci_core.c
@@ -4149,6 +4149,27 @@ int hci_recv_frame(struct hci_dev *hdev, struct sk_buff *skb)
 	/* Time stamp */
 	__net_timestamp(skb);
 
+	{
+		/* rtl8723cs bluetooth will return NOT_SUPPORTED(data[5]==1) for the WriteSecureConnectionsHostSupport (0xC7A) requests,
+		 * which will fail assert HERE: https://cs.android.com/android/platform/superproject/+/master:packages/modules/Bluetooth/system/gd/hci/controller.cc;l=225;drc=fe2308d15f93002e9dfb9a5f04d8178800ec83ea
+		 * Replace status with SUCCESS(data[5]=0) to bypass the assert. Bluetooth functionality will be limited of cause.
+	     */
+		struct hci_event_hdr *d = (void *)skb->data;
+		if (d->evt == HCI_EV_CMD_COMPLETE) {
+			struct hci_ev_cmd_complete *ec;
+			u16 opcode;
+
+			ec = (void *)(skb->data + HCI_EVENT_HDR_SIZE);
+			opcode = __le16_to_cpu(ec->opcode);
+			if (opcode == 0xC7A) {
+				printk(KERN_ERR
+				       "Recv EV CMD COMPLETE, opcode=0xC7A status code: %i, spoofing to 0 (SUCCESS)",
+				       skb->data[5]);
+				skb->data[5] = 0;
+			}
+		}
+	}
+
 	skb_queue_tail(&hdev->rx_q, skb);
 	queue_work(hdev->workqueue, &hdev->rx_work);
 
-- 
2.34.1

